odoo/tools/pdf.py

```diff
@@ -19,3 +50,258 @@ def merge_pdf(pdf_data):
+        """
+        self._attachments_data.append({
+            'filename': fname,
+            'content': fdata,
+            'subtype': fsubtype,  # optional
+        })
+
+    def write(self, stream):
+        """
+        Override from the base class to support converting to PDF/A at the moment of writing.
+        :param stream: the stream, to which we are writing the output
+        """
+        if self._is_pdfa:
+            self._convert_to_pdfa(stream)
+        else:
+            self._embed_attachments()
+            super().write(stream)
+
+    def enable_pdfa(self):
+        """
+        Mark this writer as using pdfa. This will enable the conversion to pdf/a when writing to the stream.
+        """
+        self._is_pdfa = _get_ghostscript_bin() is not False
+
+    def _convert_to_pdfa(self, stream):
+        """
+        Take the content of a pdf generated by wkhtmltopdf and use ghostscript to make PDF/A-3B compliant.
+        :return: a PDF/A-3B compliant file
+        """
+        # Replace the DOCINFO as wkhtmltopdf export one with utf-16be characters and ghostscript doesn't like it.
+        info = self.getObject(self._info)
+        info.update({
+            NameObject("/Creator"): createStringObject(""),
+            NameObject("/Title"): createStringObject("")
+        })
+        super().write(stream)
+        pdfa_content = pdf_content = stream.getvalue()
+
+        gs_bin = _get_ghostscript_bin()
+        if gs_bin:
+            try:
+                cmd = [
+                    gs_bin,
+                    '-dQUIET',  # Remove all routine information from the output
+                    '-dPDFA=3',  # Generate a PDF/A-3 file
+                    '-dBATCH',
+                    '-dNOPAUSE',
+                    '-dSAFER',  # Using DSAFER disallow access to the files on the system.
+                    '-sColorConversionStrategy=UseDeviceIndependentColor',
+                    '-sProcessColorModel=DeviceRGB',
+                    '-sDEVICE=pdfwrite',
+                    '-dPDFACompatibilityPolicy=1',  # Discard features that would result in a not PDF/A compliant file
+                    '-sOutputFile=-',  # "-" as output file makes it write to stdout instead of a file
+                    '-',  # "-" as input file is required since we provide the data from stdin
+                ]
+                # Run the command in a new subprocess. Use communicate to send the file input, and get the output
+                process = Popen(cmd, stdin=PIPE, stderr=PIPE, stdout=PIPE)
+                pdfa_content, error = process.communicate(pdf_content)
+
+                if error:
+                    _logger.error("Error while converting PDF to PDF/A: %s", error)
+                    pdfa_content = pdf_content
+            except FileNotFoundError as error:
+                _logger.error("Error while calling Ghostscript: %s", error)
+                pdfa_content = pdf_content
```

Comment:
So if pdfa is requested and we don't manage to generate it we just... ignore the entire thing?

And since we're purportedly looking up a binary beforehand where does the `FileNotFoundError` come from?

Reply:
Indeed, thank you for pointing this out. Initially, we did not use a system parameter (it was asked by Anthony recently) so we decided to silently ignore the issues since, for most users, this was not a requirement.

But as we now enable it if we want it, it would make total sense to raise errors. I did it here but still waiting for confirmation from Antony that it's the right thing to do to raise errors here.